### DISCUSSION TITLE
Low temperature on the junctions behind heat Exchangers in a DHN

### DISCUSSION THREAD

**Discussion by Kaustubh1598** [e2nIEE/pandapipes#657] [Q&A]
I am designing a DHN where i introduce an pandapipes.create_circ_pump_const_mass_flow but during computing heat transfer calulation i get negative temperature at some of the junctions. How can i implement a controller to be sure that the temperature at junction is always above ambient temp? 
![image](https://github.com/user-attachments/assets/5207872a-f99f-47ad-abb1-2f249eb3705c)
How can i tackle this issue?

**Comment by dlohmeier** [comment]
Hi @Kaustubh1598 ,
nice to see that pandapipes is used more often to model DHNs. What kind of implementation do you use for your consumers? The following suggestions from my side:
- do not use a circ_pump_const_mass, but circ_pump_const_pressure
- use the heat_consumer component 
- if you work on the develop branch (should be released soon), you can also set t_return_k for the heat consumers as the desired outlet temperature, or delta_t_k as temperature drop; then you have to use mode="bidirectional" in your pipeflow call

**Reply by Kaustubh1598** [reply]
It is a DHN with some houses and an University campus i am following the tutorial-[circular_flow_in_a_district_heating_grid.ipynb](https://github.com/e2nIEE/pandapipes/blob/master/tutorials/circular_flow_in_a_district_heating_grid.ipynb). 
For using circ_pump_const_pressure p_lift_bar value is required we have not calculated this value we have total mass flow required, the flow velocity and heat loads of the buildings in network. We use heat excahnger for every building and supply and return pipes hence a circular network
I tried using mode='bidirectional' but i get error - "No proper calculation mode chosen".
Thanks for your suggestion i will try those

**Reply by dlohmeier** [reply]
Another suggestion according to the info you have: A simple approach could be to use the heat_consumer component as consumers and there set qext_w according to the heat loads and controlled_mdot_kg_per_s to the required mass flow (i.e. mdot_hc = qext_w_hc * mdot_total / sum(heat_loads) ). But since that creates a separation between flow and return system, you need a connection somewhere, e.g. a valve at the end of a stub with extremely high loss coefficient. You should check if the mass flow via this valve gets really small and the pressure drops remain reasonable.

**Reply by Kaustubh1598** [reply]
Thanks for your important suggestion i also have these doubts-We have a Circular DHN with a constant-mass pump(We also tried Cont_Pressure pump). In this looped DHN, we have connected heat exchangers with specific flow demands. However, when we run a Pandapipes model to calculate the pressure drops and mass flows in each heat exchanger, we are getting mass flow values much lower than the ones we want. How can we ensure in Pandapipes that the required mass flow is obtained in a specific heat exchanger?
We know that the higher the pressure drop in a specific pipeline, the lower the mass flow through it. This means that higher mass flows will naturally follow pipes with lower pressure drops. What can be implemented in Pandapipes to ensure that the desired flow rate is achieved in a specific heat exchanger? Can we use a Controller?

**Comment by dlohmeier** [comment]
What you need is a flow controlling valve which is implemented in the flow_control component of pandapipes. You can put it before or after the heat_exchanger. The alternative is a heat_consumer which combines both components into one where you can set the controlled mass flow and the heat flow.

**Comment by dlohmeier** [comment]
The new heat consumer model is now released, with the possibility of a bidirectional simulation: https://github.com/e2nIEE/pandapipes/releases/tag/v0.11.0 .
